File: android/app/src/main/java/com/mewo/reader/data/LocalLibraryStore.kt
package com.mewo.reader.data

import android.content.Context
import android.graphics.Bitmap
import android.net.Uri
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.coroutines.withContext
import kotlinx.serialization.json.Json
import java.io.File
import java.util.UUID

/** On-device library: JSON metadata, EPUB files, and the feed cache. */
class LocalLibraryStore(
    private val context: Context,
    private val opener: EpubOpener,
    private val profile: LocalProfileStore,
    private val extractor: FeedExtractor = FeedExtractor(),
) : LibraryStore {
    private val json = Json {
        ignoreUnknownKeys = true
        prettyPrint = true
    }
    private val root = File(context.filesDir, "books").apply { mkdirs() }
    private val libraryFile = File(root, "library.json")
    private val mutex = Mutex()

    private val _library = MutableStateFlow(LibrarySnapshot())
    override val library: StateFlow<LibrarySnapshot> = _library.asStateFlow()

    override suspend fun load() {
        mutex.withLock {
            _library.value = readLibrary()
        }
    }

    override suspend fun importFromUri(uri: Uri): BookRecord = withContext(Dispatchers.IO) {
        val id = UUID.randomUUID().toString()
        val dir = File(root, id).apply { mkdirs() }
        val dest = File(dir, "book.epub")
        context.contentResolver.openInputStream(uri)?.use { input ->
            dest.outputStream().use { input.copyTo(it) }
        } ?: error("Could not read that file")
        ingest(id = id, file = dest, isSample = false)
    }

    override suspend fun importSample(): BookRecord = withContext(Dispatchers.IO) {
        val existing = _library.value.books.firstOrNull { it.isSample }
        if (existing != null) return@withContext existing
        val id = "sample"
        val dir = File(root, id).apply { mkdirs() }
        val dest = File(dir, "book.epub")
        context.assets.open("sample.epub").use { input ->
            dest.outputStream().use { input.copyTo(it) }
        }
        ingest(id = id, file = dest, isSample = true)
    }

    override suspend fun delete(id: String) {
        mutex.withLock {
            File(root, id).deleteRecursively()
            writeLibrary(_library.value.copy(books = _library.value.books.filterNot { it.id == id }))
        }
    }

    override suspend fun book(id: String): BookRecord? = _library.value.books.find { it.id == id }

    private fun epubFile(id: String): File = File(root, "$id/book.epub")

    override fun coverFile(id: String): File = File(root, "$id/cover.jpg")

    override suspend fun feed(id: String): List<FeedPost> = withContext(Dispatchers.IO) {
        val cacheFile = File(root, "$id/feed.json")
        if (cacheFile.exists()) {
            val cache = json.decodeFromString<FeedCache>(cacheFile.readText())
            if (cache.version == FEED_CACHE_VERSION) {
                return@withContext cache.posts
            }
        }
        val posts = opener.open(epubFile(id)) { publication ->
            extractor.extract(publication)
        }
        cacheFile.writeText(json.encodeToString(FeedCache(version = FEED_CACHE_VERSION, posts = posts)))
        mutex.withLock {
            val updated = _library.value.books.map { book ->
                if (book.id == id) {
                    book.copy(postCount = posts.size, progressIndex = 0)
                } else {
                    book
                }
            }
            writeLibrary(_library.value.copy(books = updated))
        }
        posts
    }

    override suspend fun saveProgress(id: String, index: Int) {
        mutex.withLock {
            val updated = _library.value.books.map { book ->
                if (book.id == id) book.copy(progressIndex = index.coerceAtLeast(0)) else book
            }
            writeLibrary(_library.value.copy(books = updated))
        }
    }

    override suspend fun likes(id: String): Set<String> = withContext(Dispatchers.IO) {
        val file = File(root, "$id/likes.json")
        if (!file.exists()) emptySet()
        else json.decodeFromString<LikeSet>(file.readText()).ids
    }

    override suspend fun toggleLike(id: String, postId: String): Set<String> = withContext(Dispatchers.IO) {
        val file = File(root, "$id/likes.json")
        val current = likes(id).toMutableSet()
        if (!current.add(postId)) current.remove(postId)
        file.writeText(json.encodeToString(LikeSet(current)))
        current
    }

    override suspend fun commentIndex(id: String): CommentIndex = withContext(Dispatchers.IO) {
        val file = readComments(id)
        CommentIndex(
            counts = file.posts.mapValues { it.value.size },
            mine = file.posts.filterValues { it.isNotEmpty() }.keys,
        )
    }

    override suspend fun comments(id: String, postId: String): List<PostComment> =
        withContext(Dispatchers.IO) {
            localThread(readComments(id).posts[postId].orEmpty())
        }

    override suspend fun addComment(id: String, postId: String, text: String): List<PostComment> =
        withContext(Dispatchers.IO) {
            val trimmed = text.trim()
            if (trimmed.isEmpty()) error("Write something first.")
            val asked = trimmed.length
            if (asked > MAX_COMMENT_CHARS) {
                error("Comment is longer than $MAX_COMMENT_CHARS characters (asked $asked).")
            }
            val file = File(root, "$id/comments.json")
            val current = readComments(id)
            val next = current.posts.toMutableMap()
            val row = LocalComment(
                id = UUID.randomUUID().toString(),
                text = trimmed,
                createdAt = System.currentTimeMillis(),
            )
            next[postId] = next[postId].orEmpty() + row
            file.writeText(json.encodeToString(LocalCommentFile(next)))
            localThread(next[postId].orEmpty())
        }

    override suspend fun deleteComment(
        id: String,
        postId: String,
        commentId: String,
    ): List<PostComment> = withContext(Dispatchers.IO) {
        val file = File(root, "$id/comments.json")
        val current = readComments(id)
        val next = current.posts.toMutableMap()
        next[postId] = next[postId].orEmpty().filterNot { it.id == commentId }
        if (next[postId].isNullOrEmpty()) next.remove(postId)
        file.writeText(json.encodeToString(LocalCommentFile(next)))
        localThread(next[postId].orEmpty())
    }

    private fun readComments(id: String): LocalCommentFile {
        val file = File(root, "$id/comments.json")
        if (!file.exists()) return LocalCommentFile()
        return runCatching {
            json.decodeFromString<LocalCommentFile>(file.readText())
        }.getOrDefault(LocalCommentFile())
    }

    private fun localThread(rows: List<LocalComment>): List<PostComment> {
        return rows.map { row ->
            PostComment(
                id = row.id,
                username = LOCAL_COMMENT_NAME,
                text = row.text,
                createdAt = row.createdAt,
                mine = true,
            )
        }
    }

    /**
     * Title, author, handle, plus post text from books already opened.
     * Unopened books have no feed cache, so they only match on metadata.
     */
    override suspend fun search(query: String): SearchResult = withContext(Dispatchers.IO) {
        val q = query.trim()
        if (q.isEmpty()) return@withContext SearchResult()
        if (_library.value.books.isEmpty()) {
            mutex.withLock { _library.value = readLibrary() }
        }
        val books = _library.value.books
        val bookHits = books.filter { book ->
            book.title.contains(q, ignoreCase = true) ||
                book.author.contains(q, ignoreCase = true) ||
                book.handle.contains(q, ignoreCase = true)
        }
        val postHits = books.flatMap { book ->
            val posts = cachedFeed(book.id) ?: return@flatMap emptyList()
            posts.filter { post ->
                post.text.contains(q, ignoreCase = true) ||
                    post.chapter.contains(q, ignoreCase = true)
            }.map { PostHit(book, it) }
        }
        SearchResult(books = bookHits, posts = postHits)
    }

    override suspend fun likedPosts(): List<PostHit> = withContext(Dispatchers.IO) {
        if (_library.value.books.isEmpty()) {
            mutex.withLock { _library.value = readLibrary() }
        }
        _library.value.books.flatMap { book ->
            val ids = likes(book.id)
            if (ids.isEmpty()) emptyList()
            else {
                val posts = cachedFeed(book.id) ?: emptyList()
                posts.filter { it.id in ids }.map { PostHit(book, it) }
            }
        }
    }

    override suspend fun profileReplies(handle: String): List<ProfileReply> =
        withContext(Dispatchers.IO) {
            val self = profile.profile.value.handle
            if (self.isBlank() || !handle.equals(self, ignoreCase = true)) {
                return@withContext emptyList()
            }
            if (_library.value.books.isEmpty()) {
                mutex.withLock { _library.value = readLibrary() }
            }
            _library.value.books.flatMap { book ->
                val file = readComments(book.id)
                if (file.posts.isEmpty()) return@flatMap emptyList()
                val posts = postMap(book.id)
                file.posts.flatMap { (postId, rows) ->
                    val post = posts[postId] ?: return@flatMap emptyList()
                    localThread(rows).map { comment ->
                        ProfileReply(book = book, post = post, comment = comment)
                    }
                }
            }.sortedByDescending { it.comment.createdAt }
        }

    private suspend fun postMap(id: String): Map<String, FeedPost> {
        val posts = cachedFeed(id) ?: runCatching { feed(id) }.getOrNull().orEmpty()
        return posts.associateBy { it.id }
    }

    private fun cachedFeed(id: String): List<FeedPost>? {
        val cacheFile = File(root, "$id/feed.json")
        if (!cacheFile.exists()) return null
        return runCatching {
            val cache = json.decodeFromString<FeedCache>(cacheFile.readText())
            if (cache.version == FEED_CACHE_VERSION) cache.posts else null
        }.getOrNull()
    }

    private suspend fun ingest(id: String, file: File, isSample: Boolean): BookRecord {
        val meta = opener.metadata(file)
        writeCover(id, meta.cover)
        val record = BookRecord(
            id = id,
            title = meta.title,
            author = meta.author,
            handle = slug(meta.author),
            importedAt = System.currentTimeMillis(),
            isSample = isSample,
        )
        mutex.withLock {
            val books = _library.value.books.filterNot { it.id == id } + record
            writeLibrary(LibrarySnapshot(books.sortedByDescending { it.importedAt }))
        }
        return record
    }

    private fun writeCover(id: String, cover: Bitmap?) {
        if (cover == null) return
        coverFile(id).outputStream().use { out ->
            cover.compress(Bitmap.CompressFormat.JPEG, 85, out)
        }
    }

    private fun readLibrary(): LibrarySnapshot {
        if (!libraryFile.exists()) return LibrarySnapshot()
        return runCatching {
            json.decodeFromString<LibrarySnapshot>(libraryFile.readText())
        }.getOrDefault(LibrarySnapshot())
    }

    private fun writeLibrary(snapshot: LibrarySnapshot) {
        libraryFile.writeText(json.encodeToString(snapshot))
        _library.update { snapshot }
    }
}

/** Bump when extract output changes so an already-opened book rebuilds its feed. */
const val FEED_CACHE_VERSION = 4

/** Same budget as the hosted API. See server/src/models.rs. */
const val MAX_COMMENT_CHARS = 2000

const val LOCAL_COMMENT_NAME = "You"

fun slug(value: String): String {
    val cleaned = value.lowercase()
        .replace(Regex("[^a-z0-9]+"), "")
        .take(18)
    return cleaned.ifBlank { "author" }
}

File: ios/Mewo/Data/LocalLibraryStore.swift
import Foundation
import Observation
import UIKit

/// On-device library: JSON metadata, EPUB files, and the feed cache.
@Observable
@MainActor
final class LocalLibraryStore: LibraryStore {
    private let opener: EpubOpener
    private let profile: LocalProfileStore
    private let extractor = FeedExtractor()
    private let root: URL
    private let libraryFile: URL
    private let lock = NSLock()
    private(set) var books: [BookRecord] = []

    init(opener: EpubOpener, profile: LocalProfileStore, files: URL) {
        self.opener = opener
        self.profile = profile
        self.root = files.appendingPathComponent("books", isDirectory: true)
        self.libraryFile = root.appendingPathComponent("library.json")
        try? FileManager.default.createDirectory(at: root, withIntermediateDirectories: true)
    }

    func load() async throws {
        books = readLibrary().books
    }

    func importFromFile(_ file: URL) async throws -> BookRecord {
        let id = UUID().uuidString
        let dir = root.appendingPathComponent(id, isDirectory: true)
        try FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
        let dest = dir.appendingPathComponent("book.epub")
        let accessed = file.startAccessingSecurityScopedResource()
        defer { if accessed { file.stopAccessingSecurityScopedResource() } }
        if FileManager.default.fileExists(atPath: dest.path) {
            try FileManager.default.removeItem(at: dest)
        }
        try FileManager.default.copyItem(at: file, to: dest)
        return try await ingest(id: id, file: dest, isSample: false)
    }

    func importSample() async throws -> BookRecord {
        if let existing = books.first(where: { $0.isSample }) {
            return existing
        }
        let id = "sample"
        let dir = root.appendingPathComponent(id, isDirectory: true)
        try FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
        let dest = dir.appendingPathComponent("book.epub")
        guard let bundled = Bundle.main.url(forResource: "sample", withExtension: "epub") else {
            throw NSError(domain: "mewo", code: 5, userInfo: [NSLocalizedDescriptionKey: "Sample book is missing."])
        }
        if FileManager.default.fileExists(atPath: dest.path) {
            try FileManager.default.removeItem(at: dest)
        }
        try FileManager.default.copyItem(at: bundled, to: dest)
        return try await ingest(id: id, file: dest, isSample: true)
    }

    func delete(id: String) async throws {
        try? FileManager.default.removeItem(at: root.appendingPathComponent(id, isDirectory: true))
        writeLibrary(LibrarySnapshot(books: books.filter { $0.id != id }))
    }

    func book(id: String) -> BookRecord? {
        books.first { $0.id == id }
    }

    func coverFile(id: String) -> URL? {
        let url = root.appendingPathComponent(id).appendingPathComponent("cover.jpg")
        return FileManager.default.fileExists(atPath: url.path) ? url : nil
    }

    func feed(id: String) async throws -> [FeedPost] {
        let cacheFile = root.appendingPathComponent(id).appendingPathComponent("feed.json")
        if let cache = JSONFiles.read(FeedCache.self, from: cacheFile), cache.version == feedCacheVersion {
            return cache.posts
        }
        let epub = root.appendingPathComponent(id).appendingPathComponent("book.epub")
        let posts = try await opener.open(file: epub) { publication in
            try await FeedExtractor().extract(publication)
        }
        try JSONFiles.write(FeedCache(version: feedCacheVersion, posts: posts), to: cacheFile)
        let updated = books.map { book in
            book.id == id ? BookRecord(
                id: book.id, title: book.title, author: book.author, handle: book.handle,
                importedAt: book.importedAt, postCount: posts.count, progressIndex: 0,
                isSample: book.isSample, mine: book.mine
            ) : book
        }
        writeLibrary(LibrarySnapshot(books: updated))
        return posts
    }

    func saveProgress(id: String, index: Int) async throws {
        let updated = books.map { book in
            book.id == id ? BookRecord(
                id: book.id, title: book.title, author: book.author, handle: book.handle,
                importedAt: book.importedAt, postCount: book.postCount,
                progressIndex: max(0, index), isSample: book.isSample, mine: book.mine
            ) : book
        }
        writeLibrary(LibrarySnapshot(books: updated))
    }

    func likes(id: String) async throws -> Set<String> {
        let file = root.appendingPathComponent(id).appendingPathComponent("likes.json")
        return JSONFiles.read(LikeSet.self, from: file)?.ids ?? []
    }

    func toggleLike(id: String, postId: String) async throws -> Set<String> {
        var current = try await likes(id: id)
        if !current.insert(postId).inserted {
            current.remove(postId)
        }
        let file = root.appendingPathComponent(id).appendingPathComponent("likes.json")
        try JSONFiles.write(LikeSet(ids: current), to: file)
        return current
    }

    func commentIndex(id: String) async throws -> CommentIndex {
        let file = readComments(id)
        return CommentIndex(
            counts: file.posts.mapValues(\.count),
            mine: Set(file.posts.filter { !$0.value.isEmpty }.keys)
        )
    }

    func comments(id: String, postId: String) async throws -> [PostComment] {
        localThread(readComments(id).posts[postId] ?? [])
    }

    func addComment(id: String, postId: String, text: String) async throws -> [PostComment] {
        let trimmed = text.trimmingCharacters(in: .whitespacesAndNewlines)
        if trimmed.isEmpty {
            throw NSError(domain: "mewo", code: 6, userInfo: [NSLocalizedDescriptionKey: "Write something first."])
        }
        if trimmed.count > maxCommentChars {
            throw NSError(
                domain: "mewo",
                code: 7,
                userInfo: [NSLocalizedDescriptionKey: "Comment is longer than \(maxCommentChars) characters (asked \(trimmed.count))."]
            )
        }
        var current = readComments(id)
        var next = current.posts
        let row = LocalComment(id: UUID().uuidString, text: trimmed, createdAt: nowMillis())
        next[postId, default: []].append(row)
        try JSONFiles.write(LocalCommentFile(posts: next), to: root.appendingPathComponent(id).appendingPathComponent("comments.json"))
        return localThread(next[postId] ?? [])
    }

    func deleteComment(id: String, postId: String, commentId: String) async throws -> [PostComment] {
        var current = readComments(id)
        var next = current.posts
        next[postId] = next[postId]?.filter { $0.id != commentId }
        if next[postId]?.isEmpty == true { next.removeValue(forKey: postId) }
        try JSONFiles.write(LocalCommentFile(posts: next), to: root.appendingPathComponent(id).appendingPathComponent("comments.json"))
        return localThread(next[postId] ?? [])
    }

    func search(query: String) async throws -> SearchResult {
        let q = query.trimmingCharacters(in: .whitespacesAndNewlines)
        if q.isEmpty { return SearchResult() }
        if books.isEmpty { books = readLibrary().books }
        let bookHits = books.filter {
            $0.title.localizedCaseInsensitiveContains(q)
                || $0.author.localizedCaseInsensitiveContains(q)
                || $0.handle.localizedCaseInsensitiveContains(q)
        }
        let postHits = books.flatMap { book -> [PostHit] in
            guard let posts = cachedFeed(book.id) else { return [] }
            return posts.filter {
                $0.text.localizedCaseInsensitiveContains(q) || $0.chapter.localizedCaseInsensitiveContains(q)
            }.map { PostHit(book: book, post: $0) }
        }
        return SearchResult(books: bookHits, posts: postHits)
    }

    func likedPosts() async throws -> [PostHit] {
        if books.isEmpty { books = readLibrary().books }
        var hits: [PostHit] = []
        for book in books {
            let ids = try await likes(id: book.id)
            if ids.isEmpty { continue }
            let posts = cachedFeed(book.id) ?? []
            hits.append(contentsOf: posts.filter { ids.contains($0.id) }.map { PostHit(book: book, post: $0) })
        }
        return hits
    }

    func profileReplies(handle: String) async throws -> [ProfileReply] {
        let selfHandle = await MainActor.run { profile.profile.handle }
        if selfHandle.isEmpty || handle.caseInsensitiveCompare(selfHandle) != .orderedSame {
            return []
        }
        if books.isEmpty { books = readLibrary().books }
        var replies: [ProfileReply] = []
        for book in books {
            let file = readComments(book.id)
            if file.posts.isEmpty { continue }
            let posts = await postMap(book.id)
            for (postId, rows) in file.posts {
                guard let post = posts[postId] else { continue }
                replies.append(contentsOf: localThread(rows).map {
                    ProfileReply(book: book, post: post, comment: $0)
                })
            }
        }
        return replies.sorted { $0.comment.createdAt > $1.comment.createdAt }
    }

    private func postMap(_ id: String) async -> [String: FeedPost] {
        let posts: [FeedPost]
        if let cached = cachedFeed(id) {
            posts = cached
        } else {
            posts = (try? await feed(id: id)) ?? []
        }
        return Dictionary(uniqueKeysWithValues: posts.map { ($0.id, $0) })
    }

    private func cachedFeed(_ id: String) -> [FeedPost]? {
        let cacheFile = root.appendingPathComponent(id).appendingPathComponent("feed.json")
        guard let cache = JSONFiles.read(FeedCache.self, from: cacheFile), cache.version == feedCacheVersion else {
            return nil
        }
        return cache.posts
    }

    private func ingest(id: String, file: URL, isSample: Bool) async throws -> BookRecord {
        let meta = try await opener.metadata(file: file)
        writeCover(id: id, cover: meta.cover)
        let record = BookRecord(
            id: id,
            title: meta.title,
            author: meta.author,
            handle: slug(meta.author),
            importedAt: nowMillis(),
            isSample: isSample
        )
        let next = books.filter { $0.id != id } + [record]
        writeLibrary(LibrarySnapshot(books: next.sorted { $0.importedAt > $1.importedAt }))
        return record
    }

    private func writeCover(id: String, cover: UIImage?) {
        guard let cover, let data = cover.jpegData(compressionQuality: 0.85) else { return }
        let url = root.appendingPathComponent(id).appendingPathComponent("cover.jpg")
        try? data.write(to: url, options: .atomic)
    }

    private func readComments(_ id: String) -> LocalCommentFile {
        let file = root.appendingPathComponent(id).appendingPathComponent("comments.json")
        return JSONFiles.read(LocalCommentFile.self, from: file) ?? LocalCommentFile()
    }

    private func localThread(_ rows: [LocalComment]) -> [PostComment] {
        rows.map {
            PostComment(id: $0.id, username: localCommentName, text: $0.text, createdAt: $0.createdAt, mine: true)
        }
    }

    private func readLibrary() -> LibrarySnapshot {
        JSONFiles.read(LibrarySnapshot.self, from: libraryFile) ?? LibrarySnapshot()
    }

    private func writeLibrary(_ snapshot: LibrarySnapshot) {
        try? JSONFiles.write(snapshot, to: libraryFile)
        books = snapshot.books
    }
}
